text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+    }
+}
+
+// The fields can only be set once. Calling `set_alpha` twice will result in a compilation error.
+impl<const B: bool, const G: bool> FooBuilder<false, B, G> {
+    fn set_alpha(mut self, alpha: &'static str) -> FooBuilder<true, B, G> {
+        self.alpha = Some(alpha);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+impl<const A: bool, const G: bool> FooBuilder<A, false, G> {
+    fn set_beta(mut self, beta: bool) -> FooBuilder<A, true, G> {
+        self.beta = Some(beta);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+impl<const A: bool, const B: bool> FooBuilder<A, B, false> {
+    fn set_gamma(mut self, gamma: i32) -> FooBuilder<A, B, true> {
+        self.gamma = Some(gamma);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+// If any field is optional,
+impl<const G: bool> FooBuilder<true, true, G> {
+    fn build(self) -> Foo { // can only be called if all fields have been set
+        Foo {
+            alpha: self.alpha.unwrap(),
+            beta: self.beta.unwrap(),
+            gamma: self.gamma.unwrap_or(0), // This is an optional field with a default.
+        }
+    }
+}
+
+fn main() {
+    let _ = FooBuilder::new()
+        .set_alpha("")
+        .set_beta(false) // If we comment this out, it will no longer compile.
+        .set_gamma(42) // If we comment this out, it will still compile.
+        .build();
+}
+```
+
+All of the above can be represented with the exact same results with struct
+field default values, but with much less boilerplate:
+
+```rust
+pub struct Foo {
+    pub alpha: &'static str,
+    pub beta: bool,
+    pub gamma: i32 = 0,
+}
+
+fn main() {
+    let _ = Foo {
+        alpha: "",
+        beta: false,
+        ..
+    };
+}
+```
+
+The builder pattern is quite common in the Rust ecosystem, but as shown above its need is greatly
+reduced with `struct` field defaults.
+
+## `#[derive(Default)]` in more cases
+
+The `#[derive(..)]` ("custom derive") mechanism works by defining procedural
+*macros*. Because they are macros, these operate on abstract *syntax* and
+don't have more information available. Therefore, when you `#[derive(Default)]`
+on a data type definition as with:
+
+```rust
+#[derive(Default)]
+struct Foo {
+    bar: u8,
+    baz: String,
+}
+```
+
+it only has the immediate "textual" definition available to it.
+
+Because Rust currently does not have an in-language way to define default values,
+you cannot `#[derive(Default)]` in the cases where you are not happy with the
+natural default values that each field's type provides. By extending the syntax
+of Rust such that default values can be provided, `#[derive(Default)]` can be
+used in many more circumstances and thus boilerplate is further reduced.  The
+addition of a single field, expands the code written by the `struct` author from
+a single `derive` line to a whole `Default` `impl`, which becomes more verbose
+linearly with the number of fields.
+
+### Imperfect derives
```

Comment:
@scottmcm I didn't mean this section to propose any changes to the bounds of derives, but rather argue that with default values the need for perfect default derives is lessened, no more. If removing this argument from the RFC is needed, or reword it to make that clearer, I'm more than happy to do so. I *did* file https://github.com/rust-lang/rfcs/pull/3683 to propose changes to the bounds because I felt that both changes in conjunction would work well together, but I'm not proposing changes to the bounds in *this* RFC (as they are not necessary).